File: kmeans/kmeans.py
########## kmeans.py ##########

import localsolver
import sys

if len(sys.argv) < 2:
    print("Usage: python kmeans.py inputFile [outputFile] [timeLimit] [k value]")
    sys.exit(1)

def read_elem(filename):
    with open(filename) as f:
        return [str(elem) for elem in f.read().split()]

with localsolver.LocalSolver() as ls:

    #
    # Reads instance data
    #
    if len(sys.argv) > 4: k = int(sys.argv[4])
    else: k = 2

    file_it = iter(read_elem(sys.argv[1]))
    
    # Data properties
    nb_observations = int(next(file_it))
    nb_dimensions = int(next(file_it))

    coordinates = [None]*nb_observations
    initial_clusters = [None]*nb_observations
    for o in range(nb_observations):
        coordinates[o] = [None]*(nb_dimensions)
        for d in range(nb_dimensions):
            coordinates[o][d] = float(next(file_it))
        initial_clusters[o] = next(file_it)

    #
    # Declares the optimization model
    #
    model = ls.model

    # clusters[c] represents the points in cluster c
    clusters = [model.set(nb_observations) for c in range(k)]

    # Each point must be in one cluster and one cluster only
    model.constraint(model.partition(clusters))

    # Coordinates of points
    coordinates_array = model.array(coordinates)

    # Compute variances
    variances = []
    for cluster in clusters:
        size = model.count(cluster)

        # Compute centroid of cluster
        centroid = [0 for d in range(nb_dimensions)]
        for d in range(nb_dimensions):
            coordinate_selector = model.lambda_function(
                    lambda i: model.at(coordinates_array, i, d))
            centroid[d] = model.iif(size == 0, 0,
                    model.sum(cluster, coordinate_selector) / size)

        # Compute variance of cluster
        variance = model.sum()
        for d in range(nb_dimensions):
            dimension_variance_selector = model.lambda_function(lambda i: model.sum(
                    model.pow(model.at(coordinates_array, i, d) - centroid[d], 2)))
            dimension_variance = model.sum(cluster, dimension_variance_selector)
            variance.add_operand(dimension_variance)
        variances.append(variance)

    # Minimize the total variance
    obj = model.sum(variances)
    model.minimize(obj)

    model.close()

    #
    # Parameterizes the solver
    #
    if len(sys.argv) > 3: ls.param.time_limit = int(sys.argv[3])
    else: ls.param.time_limit = 5

    ls.solve()

    #
    # Writes the solution in a file in the following format:
    #  - objective value
    #  - k
    #  - for each cluster, a line with the elements in the cluster (separated by spaces)
    #
    if len(sys.argv) > 2:
        with open(sys.argv[2], 'w') as f:
            f.write("%f\n" % obj.value)
            f.write("%d\n" % k)
            for c in range(k):
                for o in clusters[c].value:
                    f.write("%d " % o)
                f.write("\n")

File: kmeans/kmeans.cpp
//********* kmeans.cpp *********

#include <iostream>
#include <fstream>
#include <vector>
#include <limits>
#include "localsolver.h"

using namespace localsolver;
using namespace std;

class Kmeans {
public:
    // Data properties 
    int nbObservations;
    int nbDimensions;
    int k;

    vector<vector<lsdouble> > coordinates;
    vector<string> initialClusters;

    // Solver. 
    LocalSolver localsolver;

    // Decisions 
    vector<LSExpression> clusters;

    // Objective 
    LSExpression obj;

    // Constructor 
    Kmeans(int k) : k(k) { }

    // Reads instance data 
    void readInstance(const string& fileName) {
        ifstream infile;
        infile.exceptions(ifstream::failbit | ifstream::badbit);
        infile.open(fileName.c_str());

        infile >> nbObservations;
        infile >> nbDimensions;

        coordinates.resize(nbObservations);
        initialClusters.resize(nbObservations);
        for (int o = 0; o < nbObservations; o++) {
            coordinates[o].resize(nbDimensions);
            for (int d = 0; d < nbDimensions; d++) {
                infile >> coordinates[o][d];
            }
            infile >> initialClusters[o];
        }
    }

    void solve(int limit) {
        // Declares the optimization model. 
        LSModel model = localsolver.getModel();

        // Set decisions: clusters[c] represents the points in cluster c
        clusters.resize(k);
        for (int c = 0; c < k; ++c) {
            clusters[c] = model.setVar(nbObservations);
        }

        // Each point must be in one cluster and one cluster only
        model.constraint(model.partition(clusters.begin(), clusters.end()));

        // Coordinates of points
        LSExpression coordinatesArray = model.array();
        for (int o = 0; o < nbObservations; o++) {
            coordinatesArray.addOperand(model.array(coordinates[o].begin(),
                                                    coordinates[o].end()));
        }

        // Compute variances
        vector<LSExpression> variances;
        variances.resize(k);
        for (int c = 0; c < k; c++) {
            LSExpression cluster = clusters[c];
            LSExpression size = model.count(cluster);

            // Compute the centroid of the cluster
            LSExpression centroid = model.array();
            for (int d = 0; d < nbDimensions; d++) {
                LSExpression coordinateSelector = model.createLambdaFunction(
                    [&](LSExpression o) { return model.at(coordinatesArray, o, d); });
                centroid.addOperand(model.iif(size == 0, 0,
                    model.sum(cluster, coordinateSelector) / size));
            }

            // Compute the variance of the cluster
            LSExpression variance = model.sum();
            for (int d = 0; d < nbDimensions; d++) {
                LSExpression dimensionVarianceSelector = model.createLambdaFunction(
                    [&](LSExpression o) { 
                        return model.pow(model.at(coordinatesArray, o, d) 
                                         - model.at(centroid, d), 2);
                    });
                LSExpression dimensionVariance = model.sum(cluster,
                    dimensionVarianceSelector);
                variance.addOperand(dimensionVariance);
            }
            variances[c] = variance;
        }

        // Minimize the total variance
        obj = model.sum(variances.begin(), variances.end());
        model.minimize(obj);

        model.close();

        // Parameterizes the solver. 
        localsolver.getParam().setTimeLimit(limit);

        localsolver.solve(); 
    }

    // Writes the solution in a file in the following format:
    //  - objective value
    //  - k
    //  - for each cluster, a line with the elements in the cluster (separated by spaces)
    void writeSolution(const string& fileName) {
        ofstream outfile;
        outfile.exceptions(ofstream::failbit | ofstream::badbit);
        outfile.open(fileName.c_str());

        outfile << obj.getDoubleValue() << endl;
        outfile << k << endl;
        for (int c = 0; c < k; c++) {
            LSCollection clusterCollection = clusters[c].getCollectionValue();
            for (int i = 0; i < clusterCollection.count(); i++) {
                    outfile << clusterCollection[i] << " ";
                }
            outfile << endl;
        }
    }
};
    
int main(int argc, char** argv) {
    if (argc < 2) {
        cerr << "Usage: kmeans inputFile [outputFile] [timeLimit] [k value]" << endl;
        return 1;
    }

    const char* instanceFile = argv[1];
    const char* solFile = argc > 2 ? argv[2] : NULL;
    const char* strTimeLimit = argc > 3 ? argv[3] : "5";
    const char* k = argc > 4 ? argv[4] : "2";

    try {
        Kmeans model(atoi(k));
        model.readInstance(instanceFile);
        model.solve(atoi(strTimeLimit));
        if (solFile != NULL) model.writeSolution(solFile);
        return 0;
    } catch (const exception& e) {
        cerr << "An error occurred: " << e.what() << endl;
        return 1;
    }
}
